File: packages/shared/src/types/message.ts
/**
 * Shared message types for the ColorGarb order-specific message center
 */

/**
 * Message interface for order-specific communication
 */
export interface Message {
  /** Unique identifier for the message */
  id: string;
  /** ID of the user who sent this message */
  senderId: string;
  /** Display name of the message sender */
  senderName: string;
  /** Role of the message sender */
  senderRole: string;
  /** Intended recipient role for the message */
  recipientRole: string;
  /** Content of the message */
  content: string;
  /** Type/category of the message */
  messageType: MessageType;
  /** Whether the message has been read */
  isRead: boolean;
  /** Timestamp when the message was marked as read */
  readAt?: string;
  /** ID of the message this is a reply to (if applicable) */
  replyToMessageId?: string;
  /** When this message was created */
  createdAt: string;
  /** When this message was last updated */
  updatedAt: string;
  /** Number of attachments on this message */
  attachmentCount: number;
  /** List of message attachments */
  attachments: MessageAttachment[];
}

/**
 * Message attachment interface
 */
export interface MessageAttachment {
  /** Unique identifier for the attachment */
  id: string;
  /** Original filename as uploaded by the user */
  originalFileName: string;
  /** Size of the file in bytes */
  fileSize: number;
  /** MIME type of the file */
  contentType: string;
  /** When this attachment was uploaded */
  uploadedAt: string;
  /** Name of the user who uploaded this attachment */
  uploadedByName: string;
}

/**
 * Message type enum for categorization
 */
export type MessageType = 'General' | 'Question' | 'Update' | 'Urgent';

/**
 * Message search request interface
 */
export interface MessageSearchRequest {
  /** Order ID to search messages within */
  orderId: string;
  /** Search term to match against message content */
  searchTerm?: string;
  /** Filter by specific sender ID */
  senderId?: string;
  /** Filter by message type */
  messageType?: MessageType;
  /** Filter by sender role */
  senderRole?: string;
  /** Filter messages from this date (inclusive) */
  dateFrom?: string;
  /** Filter messages to this date (inclusive) */
  dateTo?: string;
  /** Include only messages with attachments */
  includeAttachments?: boolean;
  /** Page number for pagination (1-based) */
  page?: number;
  /** Number of messages per page */
  pageSize?: number;
}

/**
 * Message search response interface
 */
export interface MessageSearchResponse {
  /** Messages matching the search criteria for current page */
  messages: Message[];
  /** Total number of messages matching search criteria (before pagination) */
  totalCount: number;
  /** Current page number */
  page: number;
  /** Number of messages per page */
  pageSize: number;
  /** Whether there are more pages available */
  hasNextPage: boolean;
  /** Number of unread messages for the current user in this order */
  unreadCount: number;
}

/**
 * Send message request interface
 */
export interface SendMessageRequest {
  /** Content of the message (required, max 5000 characters) */
  content: string;
  /** Type of message (defaults to General) */
  messageType?: MessageType;
  /** Target recipient role (defaults to All) */
  recipientRole?: string;
  /** Optional ID of message being replied to */
  replyToMessageId?: string;
  /** Optional file attachments */
  attachments?: File[];
}

/**
 * Message creation response interface
 */
export interface MessageCreationResponse {
  /** The created message */
  message: Message;
  /** Any warnings about attachment processing */
  attachmentWarnings: Record<string, string>;
}

/**
 * Bulk read request interface
 */
export interface BulkReadRequest {
  /** List of message IDs to mark as read */
  messageIds: string[];
}

/**
 * Bulk read response interface
 */
export interface BulkReadResponse {
  /** Number of messages successfully marked as read */
  markedAsReadCount: number;
  /** Total number of messages requested to be marked as read */
  totalRequested: number;
}

File: apps/api/Models/Entities/Message.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace ColorGarbApi.Models.Entities;

/// <summary>
/// Represents a message in an order-specific conversation thread.
/// Provides communication channel between client organizations and ColorGarb staff.
/// </summary>
/// <since>3.3.0</since>
[Table("Messages")]
public class Message
{
    /// <summary>
    /// Unique identifier for the message
    /// </summary>
    [Key]
    public Guid Id { get; set; } = Guid.NewGuid();

    /// <summary>
    /// Reference to the order this message belongs to
    /// </summary>
    [Required]
    public Guid OrderId { get; set; }

    /// <summary>
    /// Reference to the user who sent this message
    /// </summary>
    [Required]
    public Guid SenderId { get; set; }

    /// <summary>
    /// Role of the message sender for display purposes
    /// </summary>
    [Required]
    [StringLength(20)]
    public string SenderRole { get; set; } = string.Empty;

    /// <summary>
    /// Display name of the message sender
    /// </summary>
    [Required]
    [StringLength(100)]
    public string SenderName { get; set; } = string.Empty;

    /// <summary>
    /// Intended recipient role for the message
    /// </summary>
    [Required]
    [StringLength(20)]
    public string RecipientRole { get; set; } = "All";

    /// <summary>
    /// Content of the message
    /// </summary>
    [Required]
    [StringLength(5000)]
    public string Content { get; set; } = string.Empty;

    /// <summary>
    /// Type/category of the message for prioritization and filtering
    /// </summary>
    [Required]
    [StringLength(20)]
    public string MessageType { get; set; } = "General";

    /// <summary>
    /// Indicates if the message has been read by the recipient
    /// </summary>
    public bool IsRead { get; set; } = false;

    /// <summary>
    /// Timestamp when the message was marked as read
    /// </summary>
    public DateTime? ReadAt { get; set; }

    /// <summary>
    /// Reference to the message this is a reply to (if applicable)
    /// </summary>
    public Guid? ReplyToMessageId { get; set; }

    /// <summary>
    /// When this message was created
    /// </summary>
    [Required]
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    /// <summary>
    /// When this message was last updated
    /// </summary>
    [Required]
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    /// <summary>
    /// Navigation property to the associated order
    /// </summary>
    [ForeignKey(nameof(OrderId))]
    public virtual Order Order { get; set; } = null!;

    /// <summary>
    /// Navigation property to the user who sent this message
    /// </summary>
    [ForeignKey(nameof(SenderId))]
    public virtual User Sender { get; set; } = null!;

    /// <summary>
    /// Navigation property to the parent message (if this is a reply)
    /// </summary>
    [ForeignKey(nameof(ReplyToMessageId))]
    public virtual Message? ReplyToMessage { get; set; }

    /// <summary>
    /// Navigation property for message attachments
    /// </summary>
    public virtual ICollection<MessageAttachment> Attachments { get; set; } = new List<MessageAttachment>();

    /// <summary>
    /// Navigation property for replies to this message
    /// </summary>
    [InverseProperty(nameof(ReplyToMessage))]
    public virtual ICollection<Message> Replies { get; set; } = new List<Message>();
}